pypesto/sampling/pymc3.py

```diff
@@ -0,0 +1,172 @@
+        determines one automatically (preferable).
+    **options:
+        Options are directly passed on to `pymc3.sample`.
+    """
+
+    def __init__(self, step_function=None, **options):
+        super().__init__(options)
+        self.step_function = step_function
+        self.problem: Union[Problem, None] = None
+        self.x0: Union[np.ndarray, None] = None
+        self.trace: Union[pm.backends.Text, None] = None
+        self.data: Union[az.InferenceData, None] = None
+
+    @classmethod
+    def translate_options(cls, options):
+        if options is None:
+            options = {}
+        return options
+
+    def initialize(self, problem: Problem, x0: np.ndarray):
+        self.problem = problem
+        self.x0 = x0
+        self.trace = None
+        self.data = None
+
+        self.problem.objective.history = History()
+
+    def sample(
+            self, n_samples: int, beta: float = 1.
+    ):
+        problem = self.problem
+        llh = TheanoLogLikelihood(problem, beta)
+        trace = self.trace
+
+        x0 = None
+        if self.x0 is not None:
+            x0 = {x_name: val
+                  for x_name, val in zip(self.problem.x_names, self.x0)}
+
+        # create model context
+        with pm.Model() as model:
+            # uniform prior
+            k = [pm.Uniform(x_name, lower=lb, upper=ub)
+                 for x_name, lb, ub in
+                 zip(problem.x_names, problem.lb, problem.ub)]
+
+            # convert to tensor vector
+            theta = tt.as_tensor_variable(k)
+
+            # use a DensityDist (use a lambda function to "call" the Op)
+            pm.DensityDist('llh', logp=lambda v: llh(v),
+                           observed={'v': theta})
+
+            # step, by default automatically determined by pymc3
+            step = None
+            if self.step_function:
+                step = self.step_function()
+
+            # perform the actual sampling
+            trace = pm.sample(
+                draws=int(n_samples), trace=trace, start=x0, step=step,
+                **self.options)
+
+            # convert trace to inference data object
+            data = az.from_pymc3(trace=trace, model=model)
+
+        self.trace = trace
+        self.data = data
+
+    def get_samples(self) -> McmcPtResult:
+        # parameter values
+        trace_x = np.asarray(
+            self.data.posterior.to_array()).transpose((1, 2, 0))
+
+        # TODO this is only the log-likelihood
+        trace_fval = np.asarray(self.data.log_likelihood.to_array())
+        # remove trailing dimensions
+        trace_fval = np.reshape(trace_fval, trace_fval.shape[1:-1])
+        # flip sign
+        trace_fval = - trace_fval
+
+        if trace_x.shape[0] != trace_fval.shape[0] \
+                or trace_x.shape[1] != trace_fval.shape[1] \
+                or trace_x.shape[2] != len(self.problem.x_names):
+            raise ValueError("Trace dimensions are inconsistent")
+
+        return McmcPtResult(
```

Comment:
Do we want to introduce in the results the saving of the gradient? e.g. when choosing No-U-Turn sampler

Reply:
at some point maybe useful, but not now. also, no idea how to extract it ;)